File: lims-backend/app/schemas/activity_report.py
"""Schemas — báo cáo hoạt động hàng tháng (m25). Form nhẹ, mỗi section là danh sách dòng."""
import uuid
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class TeachingEntry(BaseModel):
    course_name: str = Field(min_length=1, max_length=255)
    semester: Optional[str] = Field(default=None, max_length=32)
    hk1_theory_hours: Optional[int] = Field(default=None, ge=0, le=10000)
    hk1_practice_hours: Optional[int] = Field(default=None, ge=0, le=10000)
    hk2_theory_hours: Optional[int] = Field(default=None, ge=0, le=10000)
    hk2_practice_hours: Optional[int] = Field(default=None, ge=0, le=10000)
    note: Optional[str] = Field(default=None, max_length=2000)

    model_config = {"extra": "forbid"}


class ProjectEntry(BaseModel):
    title: str = Field(min_length=1, max_length=512)
    level: Optional[str] = Field(default=None, max_length=32)
    role: Optional[str] = Field(default=None, max_length=32)  # lead|member
    task_status: Optional[str] = Field(default=None, max_length=16)
    budget_amount: Optional[str] = Field(default=None, max_length=32)

    model_config = {"extra": "forbid"}


class PublicationEntry(BaseModel):
    pub_kind: Literal["domestic", "international", "conference"] = "domestic"
    title: str = Field(min_length=1, max_length=512)
    journal: Optional[str] = Field(default=None, max_length=255)
    year: Optional[int] = Field(default=None, ge=1900, le=2100)
    doi: Optional[str] = Field(default=None, max_length=255)
    is_scie: Optional[bool] = None
    is_ssci: Optional[bool] = None
    is_scopus: Optional[bool] = None
    is_aci: Optional[bool] = None

    model_config = {"extra": "forbid"}


class ContractEntry(BaseModel):
    title: str = Field(min_length=1, max_length=512)
    contract_type: Optional[str] = Field(default=None, max_length=64)
    value_amount: Optional[str] = Field(default=None, max_length=32)
    partner_org: Optional[str] = Field(default=None, max_length=255)

    model_config = {"extra": "forbid"}


class OtherEntry(BaseModel):
    kind: Literal["dang", "cong_doan", "vilas", "khac"] = "khac"
    content: str = Field(min_length=1, max_length=4000)

    model_config = {"extra": "forbid"}


class CreateReportRequest(BaseModel):
    period_label: str = Field(min_length=1, max_length=32)  # "01/2026"
    period_year: Optional[int] = Field(default=None, ge=1900, le=2100)
    academic_year: Optional[str] = Field(default=None, max_length=16)
    department_id: Optional[uuid.UUID] = None
    note: Optional[str] = Field(default=None, max_length=2000)
    teaching: List[TeachingEntry] = Field(default_factory=list)
    projects: List[ProjectEntry] = Field(default_factory=list)
    publications: List[PublicationEntry] = Field(default_factory=list)
    contracts: List[ContractEntry] = Field(default_factory=list)
    activities: List[OtherEntry] = Field(default_factory=list)

    model_config = {"extra": "forbid"}

File: lims-frontend/src/api/activityReport.ts
import { apiDelete, apiGet, apiGetPaged, apiPost } from '@/lib/api';
import type { ActivityReport } from '@/types';

// ── Báo cáo hoạt động hàng tháng (m25) ──────────────────────────
export interface ReportFilters {
  period?: string;
  department_id?: string;
  status?: string;
  page?: number;
  limit?: number;
}
export function listReports(f: ReportFilters = {}) {
  return apiGetPaged<ActivityReport[]>('/activity-reports', { ...f });
}
export function getReport(id: string) {
  return apiGet<ActivityReport>(`/activity-reports/${id}`);
}

// ── Payload các section ─────────────────────────────────────────
export interface TeachingEntry {
  course_name: string;
  semester?: string | null;
  hk1_theory_hours?: number | null;
  hk1_practice_hours?: number | null;
  hk2_theory_hours?: number | null;
  hk2_practice_hours?: number | null;
  hk3_theory_hours?: number | null;
  hk3_practice_hours?: number | null;
  note?: string | null;
}
export interface ProjectEntry {
  title: string;
  level?: string | null;
  role?: string | null;
  task_status?: string | null;
  budget_amount?: string | null;
}
export type PubKind = 'domestic' | 'international' | 'conference';
export interface PublicationEntry {
  pub_kind: PubKind;
  title: string;
  journal?: string | null;
  year?: number | null;
  doi?: string | null;
  is_scie?: boolean | null;
  is_ssci?: boolean | null;
  is_scopus?: boolean | null;
  is_aci?: boolean | null;
}
export interface ContractEntry {
  title: string;
  contract_type?: string | null;
  value_amount?: string | null;
  partner_org?: string | null;
}
export type OtherKind = 'dang' | 'cong_doan' | 'vilas' | 'khac';
export interface OtherEntry {
  kind: OtherKind;
  content: string;
}

export interface CreateReportBody {
  period_label: string;
  period_year?: number | null;
  academic_year?: string | null;
  department_id?: string | null;
  note?: string | null;
  teaching: TeachingEntry[];
  projects: ProjectEntry[];
  publications: PublicationEntry[];
  contracts: ContractEntry[];
  activities: OtherEntry[];
}

export function createReport(body: CreateReportBody) {
  return apiPost<ActivityReport>('/activity-reports', body);
}
export function reviewReport(id: string) {
  return apiPost<ActivityReport>(`/activity-reports/${id}/review`, {});
}
export function deleteReport(id: string) {
  return apiDelete(`/activity-reports/${id}`);
}
